### DISCUSSION TITLE
DISV equivalent for fp.mf6.utils.get_structured_faceflows

### DISCUSSION THREAD

**Discussion by byronamerson** [modflowpy/flopy#1714] [Q&A]
Hi All,

I'd like to calculate flow across a subset of the cell faces in my model using the FLOW-JA-FACE variable in the cbc file. My model consists of a DISV grid so the `get_structured_faceflows` method will not work.

Is there an alternative?  

For instance, I can pull the FLOW-JA-FACE values:
```
cbb = gwf.output.budget()
cbb.get_unique_record_names()
Qja = cbb.get_data(text='FLOW-JA-FACE')
```

And pull the values for the last stress period:
```
Qja[729]
Out[171]: 
array([[[ 0.        , -7.71035716, 11.64395461, ..., -0.        ,
          0.96678957, -0.3803444 ]]])
```
What is the structure of this data? 
Is it flow from the flow the bottom, front, and right for each cell?  So, for each stress period, total number of flow values = 3 * number of cells? 
How can I extract the values for a particular cellid?

It seems there may be a path forward with `MfGrdFile` method:

```
grb = MfGrdFile(f'{modelws}modelname.disv.grb')
modelgrid = grb.modelgrid

jas = grb.ja
```
Just not quite sure how to put it together.

Thanks!

**Comment by christianlangevin** [comment]
Hi @byronamerson, it's true that you can use the ia and ja arrays from the binary grid file to help process the flows for a DISV model.  The following excerpt is from the MODFLOW 6 input and output guide.  It shows some pseudocode for looping through the flows for individual cells.  Hopefully this will give you a possible path forward.

<img width="1095" alt="image" src="https://user-images.githubusercontent.com/3372403/217856781-23ba4768-3e8f-4196-bd47-1bc961c6ee25.png">

**Reply by jlarsen-usgs** [reply]
@byronamerson discussion #1682 has python code that can be adapted to assemble the connections between cells from `ia` and `ja`. Hopefully this helps get you started.

**Reply by byronamerson** [reply]
Thanks for the pointers @jlarsen-usgs and @langevin-usgs.  I'll review those resources.

**Reply by byronamerson** [reply]
Hello again @langevin-usgs and @jlarsen-usgs. I've spent quite a bit of time on this in the last day and I've made two conclusions:

1. I don't have a clear understanding of what `ia` and `ja` refer to in the context of identifying cells to extract flow data from, despite re-reading with mf6io.pdf and playing around with variations of the code in discussion #1682. And
2. I'm struggling to link these values with a cell id (or the info contained in `ia` and `ja`) for an arbitrary cell of interest and how to link it with the flow array from the budget file. In translating the pseudocode for the general case, I ran up against a challenge in how to use` ia`s and `ja`s to pull flow values from the `FLOWJA` array.

```
cbb = gwf.output.budget()
Qja = cbb.get_data(text = 'FLOW-JA-FACE')
Qja_one_ts = Qja[729][0:, 0:, 0:][0][0] #  FLOWJA array from one time step (I think)
grb = MfGrdFile(f'{modelws}modelname.disv.grb')

ia = grb.ia
ja = grb.ja
connections = {}
for n in range(grb.ncells):
    # print("This is cell: ", n)
    ncon = ia[n + 1] - ia[n] - 1
    if ncon < 0:
        ncon = 0
    for ipos in range(ia[n]+1,ia[n+1]):
        # print(ipos)
        m = ja[ipos]
        q = Qja_one_ts[ipos] # ipos out of range for the Q_ja_one_ts array
    print("n, m, q: ",  n, m, q)
```
It seems there is nuance in the concept of `ia` and `ja` that I just have not grasped, not to mention getting my head around the data structure of the `FLOWJA` array.

**Reply by Sakura223Leslie** [reply]
Hello. Have you found the method to solve the problem? I also have the same issue to draw the flow direction. Thank you!
